### PR TITLE
Repackaging cqlsh

### PR DESCRIPTION
**Pull request by fruch**
cqlsh is moving into it's own repository:
https://github.com/scylladb/scylla-cqlsh

* add cqlsh as submodule
* update scylla-java-tools to have cqlsh remove
* introduced new cqlsh artifact (rpm/deb/tar)

Depends: https://github.com/scylladb/scylla-tools-java/pull/316
Ref: scylladb/scylladb#11569

### CODE REVIEW

**Comment by avikivity** [comment]
I guess it failed because the tools-java dependency is broken.

Just for ci, you can repoint tools-java at your repo, and update the submodule. I will fix it up when merging.

**Comment by avikivity** [comment]
Semi-related, we can consider moving rpm/dev stuff from tools/java and tools/jmx into scylla.spec and the debian control file.

**Comment by fruch** [comment]
> I guess it failed because the tools-java dependency is broken.
> 
> Just for ci, you can repoint tools-java at your repo, and update the submodule. I will fix it up when merging.

yeah, I've forgot about that, thx

**Comment by fruch** [comment]
> Semi-related, we can consider moving rpm/dev stuff from tools/java and tools/jmx into scylla.spec and the debian control file.

also lots of the things there are repeated, all bunch of little script that needed to be duplicated (thing related to unified package for example)

**Comment by fruch** [comment]
> LGTM @fruch - is there any impact on [scylladb/scylla-pkg#2684](https://github.com/scylladb/scylla-pkg/issues/2684) with this change?

not directly, this PR just warps the cqlsh from a new repository 



I did took the name `scylla-cqlsh` 
https://pypi.org/project/scylla-cqlsh/
and we can push to it from the new repo

but it's release cycle to pypi is manual, like the drivers repositories, not coupled with scylla-core release cycle.

but let switch to scylladb/scylla-pkg#2684, since it's not relevant to the PR

**Comment by fruch** [comment]
@avikivity @yaronkaikov 

what am I doing wrong with setting the submodule to my fork ?
```
[submodule "scylla-tools"]
        path = tools/java
        url = git@github.com:fruch/scylla.git
        branch = remove_cqlsh
```

**Comment by avikivity** [comment]
Use https scheme, not ssh. I don't know if CI authenticates.

**Comment by avikivity** [comment]
```
11:18:41  fatal: remote error: upload-pack: not our ref 9420a4752a8d528cb1ef3da8cdafc7e3d86b3f16
```

Did you push that commit there?

**Comment by fruch** [comment]
> ```
> 11:18:41  fatal: remote error: upload-pack: not our ref 9420a4752a8d528cb1ef3da8cdafc7e3d86b3f16
> ```
> 
> Did you push that commit there?

yes, and tried switching to https, with no help.

**Comment by benipeled** [comment]
@fruch the submodule points to scylla repo but should points to scylla-tools-java

```
-- url = https://github.com/fruch/scylla.git
++ url = https://github.com/fruch/scylla-tools-java.git
```

**Comment by fruch** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2939/

Now need to switch scylla-pkg and scylla-dtest to work only with the unified package. instead of teaching it about the new cqlsh package.

**Comment by fruch** [comment]
@syuu1228 can you review this PR (and it's dependents)

**Comment by fruch** [comment]
@avikivity all of comment in https://github.com/scylladb/scylla-cqlsh/pull/3 we addressed.

CI is passing (including the cqlsh gating tests)

I think we are ready to move forward and kiss python2 goodbye

File: .gitmodules
Comment (yaronkaikov) [inline .gitmodules:17]: I guess you need to revert this?

Comment (fruch) [reply .gitmodules:17]: Well I think the merge who merge the tools repo, would need to remove this in sync...

**Comment by fruch** [comment]
@scylladb/scylla-maint, can some take a last look, and help me get this merged ?

**Comment by nyh** [comment]
@fruch, the commit message states: "Depends: https://github.com/scylladb/scylla-tools-java/pull/316".
Is that actually true, or that other pull request is just a cleanup which can (and even must) be done only after this one is committed?

File: .gitmodules
Comment (nyh) [reply .gitmodules:17]: @fruch I don't understand your answer. We probably don't want this setting to your private repository, not even temporarily.

**Review by nyh — CHANGES_REQUESTED** [review]
Can you please organize this patch in a way that it can be committed and only **later** will the no-longer-needed stuff be removed from the tools/java submodule? I don't want to commit it as-is with your personal github repository as a submodule...

**Comment by fruch** [comment]
> @fruch, the commit message states: "Depends: [scylladb/scylla-tools-java#316](https://github.com/scylladb/scylla-tools-java/pull/316)". Is that actually true, or that other pull request is just a cleanup which can (and even must) be done only after this one is committed?

it's must go together, otherwise the both packages create the same files in the same locations.

**Comment by fruch** [comment]
> Can you please organize this patch in a way that it can be committed and only **later** will the no-longer-needed stuff be removed from the tools/java submodule? I don't want to commit it as-is with your personal github repository as a submodule...

I had todo that in order to test it with scylla-ci in the PR, @avikivity suggest that the maintainer would take care of it (as it's need to merge with scylla-java-tools change anyhow)

you want me to remove it, that's o.k. but still this change need to go in tandem  with change in scylla-tools

**Comment by avikivity** [comment]
Please write short instructions of what needs to be done. I think I know but it's complicated enough not to trust to guesses.

**Comment by fruch** [comment]
> Please write short instructions of what needs to be done. I think I know but it's complicated enough not to trust to guesses.

@avikivity something as the following ?
- [ ] merge - https://github.com/scylladb/scylla-tools-java/pull/316
- [ ] update this PR/branch to point the master of scylla-java-tools
- [ ] merge this PR

**Comment by avikivity** [comment]
Yes. I'll try it tomorrow.

But I see cqlsh isn't in a submodule, shouldn't it be there? Otherwise an old branch can get a much newer cqlsh, and bring future regressions into that stable branch.

**Comment by fruch** [comment]
> Yes. I'll try it tomorrow.
> 
> But I see cqlsh isn't in a submodule, shouldn't it be there? Otherwise an old branch can get a much newer cqlsh, and bring future regressions into that stable branch.

it's a submodule under tool/cqlsh

I'm not sure what exactly you mean...

**Comment by avikivity** [comment]
Ah, introduced by this patch. I looked at HEAD.

**Comment by fruch** [comment]
@scylladb/scylla-maint gentle reminder

@mykaul, we do want this in 5.2, to unblock Centos/Rocky 9 (and removal of python2 in general), right ?

**Comment by nyh** [comment]
> @scylladb/scylla-maint gentle reminder
> 
> @mykaul, we do want this in 5.2, to unblock Centos/Rocky 9 (and removal of python2 in general), right ?

@avikivity please review this. I feel like I don't understand the submodule issues (I don't like @fruch's private branch solution and don't understand why it's required, but maybe you do).

**Comment by fruch** [comment]
> > @scylladb/scylla-maint gentle reminder
> > @mykaul, we do want this in 5.2, to unblock Centos/Rocky 9 (and removal of python2 in general), right ?
> 
> @avikivity please review this. I feel like I don't understand the submodule issues (I don't like @fruch's private branch solution and don't understand why it's required, but maybe you do).

It's not required for merging it was required to get scylla-ci working, this change introduce moving the cqlsh from java-tools to a new submodule of it's own, so we need both change to happen at the same time, i.e. merge the remove of cqlsh from java-tool and this one at the same time.

since I can merge it java-tools repo (yet), I was pointing to my fork, so scylla-CI can test it.

**Comment by mykaul** [comment]
> @scylladb/scylla-maint gentle reminder
> 
> @mykaul, we do want this in 5.2, to unblock Centos/Rocky 9 (and removal of python2 in general), right ?

@fruch - what are the implications of Python 2 removal on older CentOS releases (7 for example, which still should be supported).

**Comment by fruch** [comment]
> > @scylladb/scylla-maint gentle reminder
> > 
> > @mykaul, we do want this in 5.2, to unblock Centos/Rocky 9 (and removal of python2 in general), right ?
> 
> @fruch - what are the implications of Python 2 removal on older CentOS releases (7 for example, which still should be supported).

It should be using our relocate python, which is 3.11 by now (it's what is the default in fedora we are using)

I'll need to check again if we remove the python dependency all together

**Comment by avikivity** [comment]
Yes I'll review it (promised to do it Friday but...)

**Comment by avikivity** [comment]
I'd like to merge, but let's wait until the disaster around next promotion is resolved, I don't want to throw this in which can be destabilizing.

**Comment by fruch** [comment]
> I'd like to merge, but let's wait until the disaster around next promotion is resolved, I don't want to throw this in which can be destabilizing.

@avikivity are we done with the next promotion issues ?

**Comment by avikivity** [comment]
> > I'd like to merge, but let's wait until the disaster around next promotion is resolved, I don't want to throw this in which can be destabilizing.
> 
> @avikivity are we done with the next promotion issues ?

They were fixed but now resurfaced. But we can't delay it any more.

**Comment by fruch** [comment]
Found that we need to sync some changes also in scylla-pkg
- https://github.com/scylladb/scylla-pkg/pull/3286

**Comment by avikivity** [comment]
I see the scylla-pkg changes aren't merged. Will I break next promotion if I merge this?

**Comment by fruch** [comment]
> I see the scylla-pkg changes aren't merged. Will I break next promotion if I merge this?

Promotion/next won't break, but RPM/DEB installers would be broken without the scylla-pkg changes, so it better be merged in close proximity (i.e. in the same day, so the daily after that would be working)

**Comment by avikivity** [comment]
> > I see the scylla-pkg changes aren't merged. Will I break next promotion if I merge this?
> 
> Promotion/next won't break, but RPM/DEB installers would be broken without the scylla-pkg changes, so it better be merged in close proximity (i.e. in the same day, so the daily after that would be working)

Is it possible to make the scylla-pkg changes neutral, so they work both before and after? This change has a high risk of failing (it's a big change and in a risky area), so I want the ability to revert it without having to coordinate.

**Comment by fruch** [comment]
> > > I see the scylla-pkg changes aren't merged. Will I break next promotion if I merge this?
> > 
> > Promotion/next won't break, but RPM/DEB installers would be broken without the scylla-pkg changes, so it better be merged in close proximity (i.e. in the same day, so the daily after that would be working)
> 
> Is it possible to make the scylla-pkg changes neutral, so they work both before and after? This change has a high risk of failing (it's a big change and in a risky area), so I want the ability to revert it without having to coordinate.

@yaronkaikov what do you think ? beside the promotion code, it looks harmless.

I'll try to start a job testing it

**Comment by yaronkaikov** [comment]
> > > > I see the scylla-pkg changes aren't merged. Will I break next promotion if I merge this?
> > > 
> > > 
> > > Promotion/next won't break, but RPM/DEB installers would be broken without the scylla-pkg changes, so it better be merged in close proximity (i.e. in the same day, so the daily after that would be working)
> > 
> > 
> > Is it possible to make the scylla-pkg changes neutral, so they work both before and after? This change has a high risk of failing (it's a big change and in a risky area), so I want the ability to revert it without having to coordinate.
> 
> @yaronkaikov what do you think ? beside the promotion code, it looks harmless.
> 
> I'll try to start a job testing it

The changes in Scylla-pkg are related to publishing artifacts, today we fail the job if those artifacts are missing, I will go forward with testing and merging.

**Comment by fruch** [comment]
> > > I see the scylla-pkg changes aren't merged. Will I break next promotion if I merge this?
> > 
> > 
> > Promotion/next won't break, but RPM/DEB installers would be broken without the scylla-pkg changes, so it better be merged in close proximity (i.e. in the same day, so the daily after that would be working)
> 
> Is it possible to make the scylla-pkg changes neutral, so they work both before and after? This change has a high risk of failing (it's a big change and in a risky area), so I want the ability to revert it without having to coordinate.

@avikivity, so the answer is no, those changes in scylla-pkg are not backward compatible.

what else are we waiting for ?

**Comment by avikivity** [comment]
Can we make them backward compatible? Or can we test them ahead of time? I don't want to create a window where nothing works.

File: .gitmodules
Comment (avikivity) [reply .gitmodules:14]: When committing, I will change it back to ../scylla-tools-java, simultaneously merging the scylla-tools-java changes.

**Comment by avikivity** [comment]
@fruch I'm just afraid of everything stopping to work immediately after the merge because of tiny unforeseen things.

In addition to the release engineering pipeline, did you test upgrades?

**Comment by fruch** [comment]
> @fruch I'm just afraid of everything stopping to work immediately after the merge because of tiny unforeseen things.
> 
> In addition to the release engineering pipeline, did you test upgrades?

I've tested manually to see that cqlsh is working after upgrade, and that there are no conflicts while upgrading.

**Comment by avikivity** [comment]
> > @fruch I'm just afraid of everything stopping to work immediately after the merge because of tiny unforeseen things.
> > In addition to the release engineering pipeline, did you test upgrades?
> 
> I've tested manually to see that cqlsh is working after upgrade, and that there are no conflicts while upgrading.

Ok. I'll just bite the bullet.

**Comment by avikivity** [comment]
Queued. @yaronkaikov please update scylla-pkg.

**Comment by yaronkaikov** [comment]
@avikivity Merged :+1:

**Comment by fruch** [comment]
I'll stay tuned to the artifacts and upgrade  tests in the following days

**Comment by avikivity** [comment]
Amazing, it's in.

**Comment by mykaul** [comment]
> Amazing, it's in.

Great work @fruch !

**Comment by fruch** [comment]
> > Amazing, it's in.
> 
> Great work @fruch !

Thanks, still waiting for the artifacts and upgrade tests to pass on the daily build.

If it's all gonna pass without any fallout, it would be quite amazing.

**Comment by yaronkaikov** [comment]
@fruch Both `centos-rpm` and `unified-deb` passed with this new change

**Pull request closed without merging**
